homeassistant/components/sonarr/coordinator.py

```diff
@@ -0,0 +1,148 @@
+        super().__init__(
+            hass=hass,
+            logger=LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=30),
+        )
+        self.api_client = api_client
+        self.host_configuration = host_configuration
+        self.system_version: str | None = None
+
+    async def _async_update_data(self) -> T:
+        """Get the latest data from Sonarr."""
+        try:
+            return await self._fetch_data()
+
+        except exceptions.ArrConnectionException as ex:
+            raise UpdateFailed(ex) from ex
+        except exceptions.ArrAuthenticationException as ex:
+            raise ConfigEntryAuthFailed(
+                "API Key is no longer valid. Please reauthenticate"
+            ) from ex
+
+    @abstractmethod
+    async def _fetch_data(self) -> T:
+        """Fetch the actual data."""
+        raise NotImplementedError
+
+
+class CalendarDataUpdateCoordinator(SonarrDataUpdateCoordinator[list[SonarrCalendar]]):
+    """Calendar update coordinator."""
+
+    async def _fetch_data(self) -> list[SonarrCalendar]:
+        """Fetch the movies data."""
+        local = dt_util.start_of_local_day().replace(microsecond=0)
+        start = dt_util.as_utc(local)
+        end = start + timedelta(days=self.config_entry.options[CONF_UPCOMING_DAYS])
+        return cast(
+            list,
+            await self.api_client.async_get_calendar(
+                start_date=start, end_date=end, include_series=True
+            ),
+        )
+
+
+class CommandsDataUpdateCoordinator(SonarrDataUpdateCoordinator[list[Command]]):
+    """Commands update coordinator for Sonarr."""
+
+    async def _fetch_data(self) -> list[Command]:
+        """Fetch the data."""
+        return cast(list[Command], await self.api_client.async_get_commands())
+
+
+class DiskSpaceDataUpdateCoordinator(SonarrDataUpdateCoordinator[list[Diskspace]]):
+    """Disk space update coordinator for Sonarr."""
+
+    async def _fetch_data(self) -> list[Diskspace]:
+        """Fetch the data."""
+        return await self.api_client.async_get_diskspace()
+
+
+class QueueDataUpdateCoordinator(SonarrDataUpdateCoordinator[SonarrQueue]):
+    """Queue update coordinator."""
+
+    async def _fetch_data(self) -> SonarrQueue:
+        """Fetch the data."""
+        return await self.api_client.async_get_queue(
+            include_series=True, include_episode=True
+        )
+
+
+class SeriesDataUpdateCoordinator(SonarrDataUpdateCoordinator[list[SonarrSeries]]):
+    """Series update coordinator."""
+
+    async def _fetch_data(self) -> list[SonarrSeries]:
+        """Fetch the data."""
+        return cast(list[SonarrSeries], await self.api_client.async_get_series())
+
+
+class StatusDataUpdateCoordinator(SonarrDataUpdateCoordinator[str]):
+    """Status update coordinator for Sonarr."""
+
+    async def _fetch_data(self) -> str:
+        """Fetch the data."""
+        return (await self.api_client.async_get_system_status()).version
```

Comment:
I would suggest to return the whole `SystemStatus` here. The caller can then just as easily access the `version` attribute.

```suggestion
    async def _fetch_data(self) -> SystemStatus:
        """Fetch the data."""
        return await self.api_client.async_get_system_status()
```